File: 141.LinkedListCycle.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	rabbit := head
	turtle := head
	for rabbit != nil && rabbit.Next != nil {
		rabbit = rabbit.Next.Next
		turtle = turtle.Next
		if rabbit == turtle {
			return true
		}
	}
	return false
}

func main() {
	l1 := &ListNode{Val: 1}
	l2 := &ListNode{Val: 2}
	l3 := &ListNode{Val: 3}
	l4 := &ListNode{Val: 4}
	l1.Next = l2
	l2.Next = l3
	l3.Next = l4
	l4.Next = l2
	fmt.Println(hasCycle(l1))

	l1 = nil
	fmt.Println(hasCycle(l1))
}

File: linked/141.LinkedListCycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x: int) -> None:
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        rabbit = head
        turtle = head
        while rabbit is not None and rabbit.next is not None:
            rabbit = rabbit.next.next
            turtle = turtle.next
            if rabbit == turtle:
                return True
        return False


if __name__ == "__main__":
    # TestCase 1
    l1 = ListNode(3)
    l2 = ListNode(2)
    l3 = ListNode(0)
    l4 = ListNode(4)
    l1.next = l2  # type: ignore
    l2.next = l3  # type: ignore
    l3.next = l4  # type: ignore
    l4.next = l2  # type: ignore
    s = Solution()
    print(s.hasCycle(l1))

    # TestCase 2
    l1 = ListNode(1)
    l2 = ListNode(2)
    print(s.hasCycle(l1))
